Training24/FSMTest.c

```diff
@@ -0,0 +1,167 @@
+int CompareFiles (char* outputFile, char* referenceFile, int testCaseNumber);
+
+/// <summary>
+/// This function will execute the FSM providing the input and output file names as arguments
+/// </summary>
+/// <param name="exeFilePathAndName"></param>
+/// <param name="inputFilePathAndName"></param>
+/// <param name="outputFilePathAndName"></param>
+/// <returns></returns>
+int ExecProgram (char* exeFilePathAndName, char* inputFilePathAndName, char* outputFilePathAndName) {
+   char* cmdline = malloc (strlen (exeFilePathAndName) + strlen (inputFilePathAndName) + strlen (outputFilePathAndName) + 3);
+   if (cmdline == NULL) {
+      printf ("Unable to allocate memory\n");
+      return 1;
+   }
+   sprintf (cmdline, "%s%s%s%s%s", exeFilePathAndName, " ", inputFilePathAndName, " ", outputFilePathAndName);
+   // Set up structures for process information
+   STARTUPINFOA si;
+   PROCESS_INFORMATION pi;
+   // Initialize memory for STARTUPINFO and PROCESS_INFORMATION structures
+   ZeroMemory (&si, sizeof (si));
+   si.cb = sizeof (si);
+   ZeroMemory (&pi, sizeof (pi));
+   // Attempt to create the process
+   if (!CreateProcessA (
+      NULL,             // Application name (NULL if using command line)
+      cmdline,          // Command line (path to the executable)
+      NULL,            // Process handle not inheritable
+      NULL,            // Thread handle not inheritable
+      FALSE,           // Set handle inheritance to FALSE
+      0,               // No creation flags
+      NULL,            // Use parent's environment block
+      NULL,            // Use parent's starting directory
+      &si,             // Pointer to STARTUPINFO structure
+      &pi))            // Pointer to PROCESS_INFORMATION structure
+   {
+      // If CreateProcess fails, print an error message
+      int err = GetLastError ();
+      printf ("Failed to start process. Error: %lu\n", GetLastError ());
+      free (cmdline);
+      return 1;
+   }
+   // Wait until the process has finished execution
+   WaitForSingleObject (pi.hProcess, INFINITE);
+   CloseHandle (pi.hProcess);
+   CloseHandle (pi.hThread);
+   free (cmdline);
+   return 0;
+}
+
+/// <summary> Function to compare two files output and reference.</summary>
+int CompareFiles (char* outputFile, char* referenceFile, int testCaseNumber) {
+   FILE* output = fopen (outputFile, "r"), * reference = fopen (referenceFile, "r");
+   if (output == NULL || reference == NULL) {
+      printf ("Error opening files.\n");
+      return 1;
+   }
+   fseek (output, 0, SEEK_END);
+   long outputSize = ftell (output);
+   fseek (output, 0, SEEK_SET);
+   fseek (reference, 0, SEEK_END);
+   long referenceSize = ftell (reference);
+   fseek (reference, 0, SEEK_SET);
+   if (outputSize != referenceSize) {
+      printf ("Files are of different sizes. Test case %d: Expected size: %ld, Actual size: %ld\n", testCaseNumber, referenceSize, outputSize);
+      fclose (output);
+      fclose (reference);

```

Comment:
If file size is not the same, the contents of the two files are not the same. This is simply a test failure and should be reported as such.